Enable PDB archives for source-build

NikolaMilosavljevic opened the pull request:
Resolves: https://github.com/dotnet/source-build/issues/3547, https://github.com/dotnet/source-build/issues/3609 and https://github.com/dotnet/source-build/issues/3610

This change enables creation of repo-specific PDB archives, named `Symbols.<repo-name>.tar.gz`. Archives get included in `main` source-build intermediate package, and eventually extracted to `artifacts/<chip>/<flavor>` i.e. `artifacts/x64/Release`.

New infra collects all PDBs using `**\obj\**\*.pdb` pattern.

File hierarchy is preserved, which enables archive to contain multiple files with the same name. File hierarchy is rooted in `$(CurrentRepoSourceBuildArtifactsDir)` except for `source-build-externals` repo, which uses `$(RepoRoot)`.

Validation VMR build is in progress: https://dev.azure.com/dnceng/internal/_build/results?buildId=2261578&view=results

Example content, from `Symbols.sourcelink.tar.gz`:
```
obj/dotnet-sourcelink/Release/net8.0/dotnet-sourcelink.pdb
obj/Microsoft.Build.Tasks.Git/Release/net8.0/Microsoft.Build.Tasks.Git.pdb
obj/Microsoft.SourceLink.AzureDevOpsServer.Git/Release/net8.0/Microsoft.SourceLink.AzureDevOpsServer.Git.pdb
obj/Microsoft.SourceLink.AzureRepos.Git/Release/net8.0/Microsoft.SourceLink.AzureRepos.Git.pdb
obj/Microsoft.SourceLink.Bitbucket.Git/Release/net8.0/Microsoft.SourceLink.Bitbucket.Git.pdb
obj/Microsoft.SourceLink.Common/Release/net8.0/Microsoft.SourceLink.Common.pdb
obj/Microsoft.SourceLink.Gitea/Release/net8.0/Microsoft.SourceLink.Gitea.pdb
obj/Microsoft.SourceLink.Gitee/Release/net8.0/Microsoft.SourceLink.Gitee.pdb
obj/Microsoft.SourceLink.GitHub/Release/net8.0/Microsoft.SourceLink.GitHub.pdb
obj/Microsoft.SourceLink.GitLab/Release/net8.0/Microsoft.SourceLink.GitLab.pdb
obj/Microsoft.SourceLink.GitWeb/Release/net8.0/Microsoft.SourceLink.GitWeb.pdb
```

Here are some numbers:

Repo | Previous intermediate size (MB) | Symbols archive (MB) | New intermediate size (MB)
-- | -- | -- | --
arcade | 6 | 0 | 6
aspnetcore | 94 | 7 | 101
cecil | 0.6 | 0.3 | 0.9
command-line-api | 0.6 | 0.3 | 0.9
deplyment-tools | 0.1 | 0.03 | 0.13
diagnostics | 0.06 | 0.06 | 0.12
emsdk | 0.04 | 0 | 0.04
format | 14.4 | 0.07 | 14.5
fsharp | 109 | 0 | 109
installer | 203 | 0 | 203
msbuild | 8.4 | 3.4 | 11.8
nuget-client | 5.3 | 1.4 | 6.7
razor | 1 | 1 | 2
roslyn | 78 | 41 | 119
roslyn-analyzers | 18.3 | 3 | 21.3
runtime - crossgen | 256 | 0 | 256
runtime - ilcompiler | 79 | 0 | 79
runtime - main | 177 | 128 | 305
runtime - runtime | 139 | 0 | 139
sdk | 74.7 | 2.2 | 76.9
source-build-externals | 3 | 3 | 6
source-build-reference-packages | 23.5 | 0 | 23.5
sourcelink | 0.7 | 0.1 | 0.8
symreader | 0.3 | 0.06 | 0.4
templating | 1.3 | 0.6 | 2
test-templates | 0.2 | 0.004 | 0.2
vstest | 3 | 0 | 3
xdt | 0.2 | 0.1 | 0.3
xliff-tasks | 0.1 | 0.04 | 0.15
Totals | 1297 | 192 | 1489

mthalman commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildIntermediate.proj:
```diff
@@ -79,6 +81,48 @@
   </Target>
 
+  <!-- Discover all repo's symbols that match **\obj\**\*.pdb, to be included in the "main" intermediate nupkg. -->
+  <Target Name="CreateSymbolsList" Condition="'$(SupplementalIntermediateNupkgCategory)' == ''">
```
While I appreciate the breaking down of logic into smaller chunks, it feels weird to me that `CreateSymbolsList` is a separate target whose output is then deleted by another target: `CreateSymbolsArchive`. It seems like the symbols list is just an implementation detail of the logic to create the archive and doesn't deserve a separate target.

NikolaMilosavljevic replied:
Yes, this is true. It started as a single target which I split into two to simplify conditioning of few tasks (packaging, message, deletion of symbols list) - each task would need its own condition.

However, I see the point and will revert to my original design.

NikolaMilosavljevic replied:
Fixed with https://github.com/dotnet/arcade/pull/14015/commits/de9bb57fb47d9e329ee161128f0846aad636b215

NikolaMilosavljevic commented:
Found a small issue in one of the conditions - it caused few issues: 1) no archive for `source-build-externals` and 2) unnecessarily longer hierarchy in most packages. Testing the fix locally...

omajid commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildIntermediate.proj:
```diff
@@ -79,8 +80,44 @@
+    </PropertyGroup>
+
+    <ItemGroup>
+      <AbsoluteSymbolPath Include="$(SymbolsRoot)\**\obj\**\*.pdb" />
```
Apologies if this is a stupid question, but is there any risk of a not-just-built pdb file being available at this path? Is there any support for posioning pdb files to catch potential leaks?

NikolaMilosavljevic replied:
`obj` folder would contain content of just compiled sources. `bin` folder would also contain dependencies.

MichaelSimons commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildIntermediate.proj:
```diff
@@ -79,8 +80,44 @@
   </Target>
 
+  <!-- Create symbols archive and include it in the "main" intermediate nupkg. -->
+  <Target Name="GetSymbolsArchive" Condition="'$(SupplementalIntermediateNupkgCategory)' == ''">
```
Does this gather any symbols for runtime?  Wondering how this works with their existing symbols tarball.  Is there a risk that there are multiple runtime symbol tarballs?

NikolaMilosavljevic replied:
Yes, it does. Main intermediate package does not have `SupplementalIntermediateNupkgCategory` property set. I have verified contents of symbols packages for all repos.

MichaelSimons commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildIntermediate.proj:
```diff
@@ -79,8 +80,44 @@
+
+    <ItemGroup>
+      <AbsoluteSymbolPath Include="$(SymbolsRoot)\**\obj\**\*.pdb" />
+      <RelativeSymbolPath Include="@(AbsoluteSymbolPath)" Condition="'@(AbsoluteSymbolPath)' != ''">
```
Curious if this can be achieved in one ItemGroup

<Symbols Include="$(SymbolsRoot)\**\obj\**\*.pdb" >
    <RelativePath>$([MSBuild]::MakeRelative($(SymbolsRoot), %(FullPath)))</RelativePath>
</Symbols>

NikolaMilosavljevic replied:
Will test it out - thanks!

NikolaMilosavljevic replied:
We can do something like this:
```
      <AbsoluteSymbolPath Include="$(SymbolsRoot)\**\obj\**\*.pdb" />
      <AbsoluteSymbolPath Update="@(AbsoluteSymbolPath)" Condition="'@(AbsoluteSymbolPath)' != ''">
        <RelativePath>$([MSBuild]::MakeRelative($(SymbolsRoot), %(FullPath)))</RelativePath>
      </AbsoluteSymbolPath>
```

It removes one ItemGroup.

NikolaMilosavljevic replied:
Fixed with https://github.com/dotnet/arcade/pull/14015/commits/e41ac2feebf07917f3afa42811d83e3646714cdd

MichaelSimons commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildIntermediate.proj:
```diff
@@ -79,8 +80,44 @@
+    </PropertyGroup>
+
+    <ItemGroup>
+      <AbsoluteSymbolPath Include="$(SymbolsRoot)\**\obj\**\*.pdb" />
```
I typically seem ItemGroups name in plural form as the are a type of collection.  e.g. AbsoluteSymbolPaths

NikolaMilosavljevic replied:
I've seen both plural and singular in use and can't seem to find any guidance for .NET codebase. It appears that even source-build infra is inconsistent, i.e.

**Singular**:
https://github.com/dotnet/arcade/blob/08a46d5d5fb05f3d55de3955535996a630d75e2a/src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcade.targets#L83
https://github.com/dotnet/arcade/blob/08a46d5d5fb05f3d55de3955535996a630d75e2a/src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildIntermediate.proj#L60

**Plural**:
https://github.com/dotnet/arcade/blob/08a46d5d5fb05f3d55de3955535996a630d75e2a/src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildArcadeBuild.targets#L115
https://github.com/dotnet/arcade/blob/08a46d5d5fb05f3d55de3955535996a630d75e2a/src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/AfterSourceBuild.proj#L23

MichaelSimons replied:
I acknowledge there are inconsistencies in the code base.  I asked to see if you had reasons for using singular.  I personally find the plural form can help the readability therefore tend to favor that form as a general rule.

MichaelSimons reviewed (commented):
Looks good.  I'd be curious to see what the size impact is to the current intermediates.  I know runtime has run into the 500MB limit AZDO has on individual artifact files.

NikolaMilosavljevic commented:
> Looks good. I'd be curious to see what the size impact is to the current intermediates. I know runtime has run into the 500MB limit AZDO has on individual artifact files.

Added the relevant data to main comment.

NikolaMilosavljevic commented on src/Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/SourceBuildIntermediate.proj:
```diff
@@ -79,8 +80,44 @@
+      <SymbolsArchiveLocation>$(CurrentRepoSourceBuildArtifactsPackagesDir)</SymbolsArchiveLocation>
+      <SymbolsList>$(SymbolsArchiveLocation)\symbols.lst</SymbolsList>
+      <SymbolsArchivePrefix>Symbols.</SymbolsArchivePrefix>
+      <SymbolsArchiveName>$(SymbolsArchiveLocation)$(SymbolsArchivePrefix)$(GitHubRepositoryName).tar.gz</SymbolsArchiveName>
```
That's a good idea - will update the name so it matches other artifacts.

NikolaMilosavljevic replied:
It seems that only `runtime` and `sdk` have non-portable RID set in `TargetRid` property. Other repos use the portable Rid, as we are doing portable source-builds for majority of the repos..

My test build with both `version` and `rid` produced the following - naming is inconsistent, but it matches the Rid used for repo's build:
```
Symbols.aspnetcore.8.0.0-rc.2.23423.7.linux-x64.tar.gz
Symbols.cecil.0.11.4.0-alpha.23407.2.linux-x64.tar.gz
Symbols.command-line-api.0.1.435601.linux-x64.tar.gz
Symbols.deployment-tools.8.0.0-preview.6.23407.1.linux-x64.tar.gz
Symbols.diagnostics.7.0.0-preview.23211.1.linux-x64.tar.gz
Symbols.format.8.0.442201.linux-x64.tar.gz
Symbols.msbuild.17.8.0-preview-23424-01.linux-x64.tar.gz
Symbols.nuget-client.1.0.0-preview.23424.1.linux-x64.tar.gz
Symbols.razor.7.0.0-preview.23423.3.linux-x64.tar.gz
Symbols.roslyn-analyzers.3.11.0-beta1.23420.2.linux-x64.tar.gz
Symbols.roslyn.4.8.0-2.23423.10.linux-x64.tar.gz
Symbols.runtime.8.0.0-rc.2.23418.14.fedora.36-x64.tar.gz
Symbols.sdk.8.0.100-rc.2.23423.14.fedora.36-x64.tar.gz
Symbols.source-build-externals.8.0.0-alpha.1.23421.1.linux-x64.tar.gz
Symbols.sourcelink.8.0.0-beta.23422.2.linux-x64.tar.gz
Symbols.symreader.2.1.0-beta.23253.1.linux-x64.tar.gz
Symbols.templating.8.0.100-rc.2.23423.2.linux-x64.tar.gz
Symbols.test-templates.1.1.0-rc.23410.2.linux-x64.tar.gz
Symbols.xdt.7.0.0-preview.22423.2.linux-x64.tar.gz
Symbols.xliff-tasks.1.0.0-beta.23418.1.linux-x64.tar.gz
```

NikolaMilosavljevic replied:
Fixed with https://github.com/dotnet/arcade/pull/14015/commits/e41ac2feebf07917f3afa42811d83e3646714cdd

NikolaMilosavljevic commented:
> Found a small issue in one of the conditions - it caused few issues: 1) no archive for `source-build-externals` and 2) unnecessarily longer hierarchy in most packages. Testing the fix locally...

This is now fixed. I've updated the initial comment with updated symbols paths in archive packages and added all relevant numbers for intermediate and symbols archive sizes.

mthalman reviewed (dismissed):
We want this for .NET 8, right? It's easier to just target release/8.0.1xx first and, once it's merged, it will then automatically be ported to main branch.

NikolaMilosavljevic commented:
> We want this for .NET 8, right? It's easier to just target release/8.0.1xx first and, once it's merged, it will then automatically be ported to main branch.

Arcade repo doesn't have any 8.0-specific branch, they flow their `main` branch to `installer`'s `release/8.0.1xx` branch; e.g. https://github.com/dotnet/installer/pull/17293

I'm going to make a small update to this PR to address 2 of @MichaelSimons comments.
